### ISSUE TITLE
Custom response handling

### ISSUE THREAD

**Issue by sirudog** [arangodb/arangojs#396]
Hi,

I use a custom JSON serializer for my entities which are persisted to ArangoDB via a Foxx service.
This works fine, I just add the serialized entity as string to `async route.post([path,] [body, [qs]]): Response` method (Route API). arangojs realizes that the body is already string so it keeps out of my way and just sends it to my Foxx service which saves it into my entities collection.

What does not work is retrieving these entities. 

The main reason is that I cannot hook into the deserialization pipeline of arangojs, meaning the response.body is always deserialized by JSON.parse if the mime type is set to "application/json":
https://github.com/arangodb/arangojs/blob/5387a6e560418380a99fc6dc3c59b433b2570bdf/src/connection.js#L119-L127

Luckily I have a workaround, by manually setting the MIME type to "text/plain" and JSON.stringifying the result set converted to an array in my Foxx Service. This way arangojs does not touch the response body.

What I would like to do is to get the raw body as is, even if the mime type is "application/json", so I can use my JSON serializer to deserialize it properly into my entity type. 
Is this scenario worth to look into?

**Comment by pluma** [comment]
Have you tried passing the option `expectBinary`? This should bypass the response parsing except for errors. You'd have to use the `request` method directly instead of the wrappers but it should restore the raw body for non-error responses.

**Comment by pluma** [comment]
FWIW, it's not necessary to use arangojs if you only want to talk to a Foxx service. Any conventional HTTP client will do. It's only recommended to use arangojs to talk to Foxx services if you already use arangojs to talk to the ArangoDB HTTP API (e.g. to manage collections).

**Comment by sirudog** [comment]
Thanks for the hint about `expectBinary` option. I will try it as using it sounds better than tampering with my Foxx service. I would still stick to arangojs, as I like to be as close to ArangoDB as possible, especially because I am looking forward to using VelocyStream... It seems to me that it is already on the way?
I will report back how did it go, maybe other will find the outcome useful, too

**Comment by pluma** [comment]
Yes, we're currently working on porting VelocyStream to Node via the Fuerte project.

**Comment by sirudog** [comment]
Thanks for the info about VelocyStream.

I have checked the `request` method (in route.js), but it seems to me that `expectBinary` option would not solve my problem. Actually what more it would if I set `expectBinary` true do is just calling `toString('utf-8')` on the res.body, but in the next line it would still `JSON.parse` the res.body, since that line is not in the if block. 
Please correct me if I am wrong. Thanks
```
if (res.headers['content-type'].match(MIME_JSON)) {
            try {
=>            if (expectBinary) res.body = res.body.toString('utf-8')  // THIS LINE WOULD NOW BE EXECUTED
=>            res.body = res.body ? JSON.parse(res.body) : undefined // THIS LINE IS ALWAYS EXECUTED
            } catch (e) {
              res.body = rawBody
              e.response = res
              return callback(e)
            }
```

**Comment by pluma** [comment]
That is correct. It will try to parse the body and it will try to detect ArangoDB error responses but it will reset the `res.body` for non-error responses before invoking the callback.

So this should have you covered except in two situations:

1. You're sending malformed JSON data (i.e. `JSON.parse` will throw a syntax error) with the JSON mime type
2. Your error responses need special logic as well.

**Comment by sirudog** [comment]
OK, I revisited this issue and it would indeed reset the `res.body` as you explained. 
The reason why I still would not use the `expectBinary` workaround is that the response body is unnecessarily `JSON.parse`d before this reset happens. So I decided to keep the MIME type workaround in my Foxx Service as it was already in place.

Thanks for your help, I really enjoy working with ArangoDB. You guys doing a terrific job!
